### ISSUE TITLE
Determine how to store history

### ISSUE THREAD

**Issue by ArendPeter** [Equal-Vote/star-server#135]
TLDR: I like the **Single Audit Table** design. Most of the sources I've read seem to recommend **Shadow History Tables**, but the format of the audit table feels like a perfect fit for the information we want to display, and I'm not sure most of the performance warnings apply to us (please double check me on this)

(originally thread [here](https://github.com/Equal-Vote/star-server/pull/123#discussion_r928385075) , but I wanted to make an issue for it)

To expand our conclusions from the thread, I think we can break it down into these strategies

## Decentralized Blobs
At the moment we have separate history information for voter rolls, ballots, and elections, and these are each stored in their corresponding tables as a JSON blob
 🔴**Auditing**: We'd have to combine a bunch of lists to create an auditor view
 🟢**Data Contention**: Essentially no data contention (Data contention issues will be very rare as it will require multiple people modifying the same row)
 🟢**Common Case Performance**: Auditing is not the common case (ex. if it takes an hour to generate the audit log, that's probably fine). User queries (viewing history on a single ballot, an admin viewing recent election updates, etc) are the common case, and this design optimizes for this
🟢**Effort**: This let's us continue with our current database design, and the history is well suited for common case queries

## 1 Blob per election
Here each election has an entry for it's history and we update the blob whenever something happens in any table
🟢**Auditing**: Auditing would be a single query on the history table
🔴**Data Contention**: All the data in one place means there's more data contention
🔴**Common case performance**: Anytime a user views a ballot or an admin view recent updates, the request will need to access the history blob, and filter it. That's lots of requests going to the same place
🔴**Effort**: We'd need to add a lot of logic for the common case queries to deconstruct and filter the blob

Recommend Reading on the audit pattern vs the history pattern: https://dev.to/zhiyueyi/design-a-table-to-keep-historical-changes-in-database-10fn

## Shadow History Tables
Here we create a shadow history table for each table we care about. The history table has an entry for every update that happened to the primary table
🟡**Auditing**: The events are spread out across multiple tables, but it's still better than the Decentralized blob approach because you can do mysql joins rather than implement JSON merging logic
🟢**Data Contention**: Adding history rows won't cause any data contention (however we still need to have safety in place for multiple people modifying the same row)
🟡**Common case performance**: Anytime a user views a ballot or an admin view recent updates, the request will need to do a filter query on the corresponding shadow history table. Not as clean as the Decentralized Blob, but still ok
🟡**Effort**: We'd need to create additional tables and add the corresponding code to account for them. We can use triggers to avoid needing to create duplicate insertion commands everywhere

## Single Audit Table
Here updates to ANY table will be logged into a single audit table. Unlike history tables, audit tables are setup to be generic so all updates will be compatible with it
🟢**Auditing**: No need to merge tables, we can generate the audit csv with a single filtered query to the audit table
🟢**Data Contention**: Add audit table rows won't cause any data contention (however we still need to have safety in place for multiple people modifying the same row)
🟡**Common case performance**:  Pretty similar to the (Shadow History Tables) solution. However one benefit is the audit table format can better show the exact change that was made. We can also centralize an "audit rendered" class which converts audit table entries into human readable text. This text can be reused whenever we're displaying history
🟡**Effort**: We'd need to create additional tables and add the corresponding code to account for them. We can use triggers to avoid needing to create duplicate insertion commands everywhere

> WARNING: A lot of people in the reference links have problems with the performance of the Audit Table pattern (see [example](https://stackoverflow.com/questions/7975317/audit-tables-each-field-for-table-or-one-table) ), but I don't think most of them apply to us? The columns need to be nvarchar(max), so that could cause space issues, but I believe our max won't be too high. Generating point in time snapshots can be hard with the audit table, but I don't see us needing that. 

## References
https://www.codeproject.com/Articles/105768/Audit-Trail-Tracing-Data-Changes-in-Database
https://stackoverflow.com/questions/12563706/is-there-a-mysql-option-feature-to-track-history-of-changes-to-records
https://vertabelo.com/blog/database-design-for-audit-logging/
https://stackoverflow.com/questions/7975317/audit-tables-each-field-for-table-or-one-table
https://stackoverflow.com/questions/16881043/are-database-triggers-safe-for-cross-table-integrity-constraints

**Comment by ArendPeter** [comment]
This article seems relevant https://yizhiyue.me/2019/09/17/design-a-table-to-keep-historical-changes-in-database

Their approaches go into how to model seperate history / auditor tables, but didn't cover our current json blob approach. I left a comment but the article's 3 years old so who knows if they'll see it

**Comment by scottPlusPlus** [comment]
I am not tied to the decentralized blobs, but current situation seems to work fine now so would prefer to spend effort on other issues.

1. We only need to generate an entire-election audit once, so not worried about performance
2. We would need to update all our writes to use transactional writes across two tables

**Comment by ArendPeter** [comment]
> I am not tied to the decentralized blobs, but current situation seems to work fine now so would prefer to spend effort on other issues.
Yep, that's fair

> 1. We only need to generate an entire-election audit once, so not worried about performance
Also fair

> 2. We would need to update all our writes to use transactional writes across two tables
I was thinking we use database triggers here. So our code only interacts with the primary database, and then those updates trigger a secondary flow to update the audit table

**Comment by scottPlusPlus** [comment]
> I was thinking we use database triggers here. So our code only interacts with the primary database, and then those updates trigger a secondary flow to update the audit table

Awesome.  I've been thinking about converting the whole system to use event-driven architecture (if resources were not an issue), but will take a look at these update triggers.

**Comment by ArendPeter** [comment]
We've recently redesigned our data scheme to make it easier to track version history.  It'll still require combining tables for the purposes of auditing but as Scott said, this won't need to be done super often so it's not too much of an issue
